src/modules/flow/ipm/ipm.c

```diff
@@ -0,0 +1,599 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2015 Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "sol-flow/ipm.h"
+#include "sol-flow-internal.h"
+#include "sol-ipm.h"
+
+#include <sol-util.h>
+#include <errno.h>
+
+/* TODO would be lovely to have this available globally... create a new
+* blob from some variable - but duplicating it, not just wrapping it */
+#define CREATE_BLOB(_val) \
+    struct sol_blob *blob; \
+    size_t _s = sizeof(*_val); \
+    void *v = malloc(_s); \
+    SOL_NULL_CHECK(v, -ENOMEM); \
+    blob = sol_blob_new(SOL_BLOB_TYPE_DEFAULT, NULL, v, _s); \
+    if (!blob) { \
+        free(v); \
+        return -EINVAL; \
+    }
+
+#define UPDATE_BLOB(_blob, _val) \
+    memcpy(_blob->mem, _val, sizeof(*_val));
+
+
+#define SEND_BLOB(_blob, _id, _node) \
+    do { \
+        r = sol_ipm_send(_id, _blob, NULL, NULL); \
+        if (r < 0) \
+            sol_flow_send_error_packet(_node, -r, "Could not send IPM message"); \
+        sol_blob_unref(_blob); \
+    } while (0)
+
+struct ipm_data {
+    uint32_t id;
+};
+
+static void
+common_reader_close(struct sol_flow_node *node, void *data)
+{
+    struct ipm_data *mdata = data;
+
+    sol_ipm_set_receiver(mdata->id, NULL, NULL);
+}
+
+/*********************** Boolean node *****************************/
+
+static int
+boolean_writer_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    struct ipm_data *mdata = data;
+    const struct sol_flow_node_type_ipm_boolean_writer_options *opts;
+
+    SOL_FLOW_NODE_OPTIONS_SUB_API_CHECK(options, SOL_FLOW_NODE_TYPE_IPM_BOOLEAN_WRITER_OPTIONS_API_VERSION,
+        -EINVAL);
+    opts = (const struct sol_flow_node_type_ipm_boolean_writer_options *)options;
+
+    SOL_INT_CHECK(opts->id, < 1, -EINVAL);
+    SOL_INT_CHECK(opts->id, > sol_ipm_max_id(), -EINVAL);
+
+    mdata->id = opts->id;
+
+    return 0;
+}
+
+static int
+boolean_writer_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    struct ipm_data *mdata = data;
+    bool in_value;
+    int r;
+
+    CREATE_BLOB(&in_value);
+
+    r = sol_flow_packet_get_boolean(packet, &in_value);
+    SOL_INT_CHECK(r, < 0, r);
+
+    UPDATE_BLOB(blob, &in_value);
```

Comment:
as @glima said, make the wrapper work on the final value. Then you declare `struct sol_blob *blob;` yourself and call: `blob = SOL_BLOB_NEW_DUP(in_value)`, with:

```c
static inline struct sol_blob *
sol_blob_new_dup(const void *mem, size_t size) {
    ...;
}
#define SOL_BLOB_NEW_DUP(x) sol_blob_new_dup((x), sizeof(x))
```